Improve Chat Input with Auto-Sizing Textarea

characharm opened the pull request:
This PR proposes an enhancement to the chat input `textarea` in the web UI.

**Problem:**

Currently, the chat input textarea has a fixed initial height and relies on manual resizing via the handle in the bottom-right corner. This manual resizing can sometimes feel unintuitive, especially as the sticky-positioned input area visually expands upwards when resized from the bottom. Furthermore, composing longer, multi-line messages often requires frequent manual resizing or scrolling within a constrained areea.

**Proposed Solution:**

This change replaces the manual resizing behavior with an **auto-sizing** mechanism:

1.  The `textarea` now automatically adjusts its height vertically to fit the content as the user types or pastes text.
2.  A maximum height constraint (`max-h-48` / `12rem` by default in the current implementation) is applied.
3.  Once the content exceeds this maximum height, a vertical scrollbar appears *inside* the `textarea`, preventing the input area from growing indefinitely.
4.  The native browser resize handle has been disabled (`resize: none`).

**Implementation Details:**

*   A new custom React hook, `useAutosizeTextarea`, has been created to encapsulate the logic.
*   This hook utilizes a `ref` to directly interact with the `textarea` DOM element.
*   On the `onInput` event, the hook calculates the `scrollHeight` of the `textarea` and updates its inline `height` style accordingly, respecting the `max-height` set via CSS.
*   This implementation **preserves the existing optimization** of avoiding component re-renders on every keystroke 

**Benefits:**

*   Provides a more fluid and intuitive user experience for writing multi-line messages.
*   Eliminates the need for manual resizing for typical message lengths.
*   Offers predictable behavior by limiting excessive growth with `max-height`.

---

**Note:**

Please consider this pull request more as a **suggestion** 
master:

https://github.com/user-attachments/assets/51567ee9-f6dc-423e-ae07-8efa4d7e6b73

pr: 

https://github.com/user-attachments/assets/a0aa972b-d897-42bb-9dba-a6a4a5a9d0c6

characharm committed:
Update ChatScreen.tsx

characharm committed:
useAutosizeTextarea.ts

useAutosizeTextarea to encapsulate the logic.

ngxson commented on examples/server/webui/src/components/ChatScreen.tsx:
```diff
@@ -286,43 +290,3 @@ export default function ChatScreen() {
-  ref: React.RefObject<HTMLTextAreaElement>;
-}
-
-// This is a workaround to prevent the textarea from re-rendering when the inner content changes
-// See https://github.com/ggml-org/llama.cpp/pull/12299
```
We should preserve these comments (and you should also add comments to explain your code if necessary)

ngxson commented on examples/server/webui/src/components/useAutosizeTextarea.ts:
```diff
@@ -0,0 +1,68 @@
+  onInput: (event: React.FormEvent<HTMLTextAreaElement>) => void;  
+}
+
+export function useAutosizeTextarea(initValue: string): AutosizeTextareaApi {
```
Now that this hook has 2 purposes (auto-resize + optimize rendering), I think we should have a more generic name for it, maybe `useChatTextarea`

ngxson reviewed (commented):
The overall idea is good, just need a bit of clean up. Also, don't forget to run `npm run build`

ngxson commented:
Also, I'm thinking if we should disable this on mobile view. Otherwise, it will take up a large space on small screen

characharm committed:
Implement responsive auto-sizing chat textarea

Replaces the manual textarea resizing with an automatic height adjustment based on content.

- `useChatTextarea` hook to manage textarea state and auto-sizing logic via refs, preserving the optimization
- Textarea now grows vertically up to a maximum height (`lg:max-h-48`) on large screens (lg breakpoint and up).
- Disables auto-sizing and enables manual vertical resizing (`resize-vertical`) on smaller screens for better mobile usability.
- Aligns the "Send" button to the bottom of the textarea (`items-end`) for consistent positioning during resize.

characharm committed:
-update compressed index.html.gz after npm run build
-refactor: replace OptimizedTextareaValue with AutosizeTextareaApi in VSCode context hook

characharm committed:
chore: normalize line endings to LF
refactor: AutosizeTextareaApi -> chatTextareaApi

characharm committed:
refactor: Rename interface to PascalCase

ngxson reviewed (approved):
Thanks for adding comments, it makes the code much easier to understand!

ngxson committed:
Merge branch 'master' into useAutosizeTextarea

woof-dog commented:
This change unfortunately results in catastrophic and devastating huge lag when you paste a giant string into the textarea. Full CPU, every keystroke takes 2 seconds or longer.

ngxson commented:
@woof-dog the laggy behavior should be addressed in https://github.com/ggml-org/llama.cpp/pull/13365/commits/163bdca511be3e229e18197ee4e1dfdabd53991c
